### PR TITLE
Fix "log.SetLogger(...) was never called"

### PR DESCRIPTION
**Pull request by sheidkamp**
# Description

Set a logger in the `BeforeSuite` of the gloo kubes e2e test to prevent the condition leading to the display of the error message.

Be sure to verify intended behavior by inspecting CI log runs and/or running the e2e tests manually yourself.

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/8548

### CODE REVIEW

**Review by elcasteel — DISMISSED** [review]
Does this need to happen in any other tests?

**Review by sam-heilbron — COMMENTED** [review]
In solo-projects we initialize the logger (https://github.com/solo-io/solo-projects/blob/36772c84eb1522d64581404fe58a02a5865a1a87/projects/gloo/pkg/bootstrap/context.go#L70). 

What's interesting is that this stack trace is explicitly for our RateLimitConfigClients (https://github.com/solo-io/gloo/blob/main/projects/gloo/pkg/api/external/solo/ratelimit/extensions.go#L65). The RateLimit API is distinct from other APIs in that it is built using skv2 (instead of solo-kit) and therefore relies on the controller-runtime library more broadly.

Do we know why invoking the RateLimitConfigClients is where we hit this? I wonder if what we really need is to set the logger (like we do in the solo-projects code I linked) in the setup to the gloo component.

File: test/kube2e/gloo/gloo_suite_test.go [hunk 1]
```diff
@@ -51,6 +53,8 @@ var (
 var _ = BeforeSuite(func() {
 	var err error
 
+	// This line prevents controller-runtime from complaining about log.SetLogger never being called
```
[the remark this answers is not in the document]
Comment (sheidkamp) [reply test/kube2e/gloo/gloo_suite_test.go:56]: I looked through the logs in CI and only found the error in this test suite.

File: changelog/v1.15.0-rc4/set-logger-in-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+changelog:
```
[the remark this answers is not in the document]
Comment (sheidkamp) [reply changelog/v1.15.0-rc4/set-logger-in-tests.yaml]: moved

**Comment by sheidkamp** [comment]
> In solo-projects we initialize the logger (https://github.com/solo-io/solo-projects/blob/36772c84eb1522d64581404fe58a02a5865a1a87/projects/gloo/pkg/bootstrap/context.go#L70).
> 
> What's interesting is that this stack trace is explicitly for our RateLimitConfigClients (https://github.com/solo-io/gloo/blob/main/projects/gloo/pkg/api/external/solo/ratelimit/extensions.go#L65). The RateLimit API is distinct from other APIs in that it is built using skv2 (instead of solo-kit) and therefore relies on the controller-runtime library more broadly.
> 
> Do we know why invoking the RateLimitConfigClients is where we hit this? I wonder if what we really need is to set the logger (like we do in the solo-projects code I linked) in the setup to the gloo component.

RateLimitConfigClients is where we hit this because the other "getClient" calls are to modules that use solo kit instead of importing the `controller-runtime` client directly.

Are you suggesting we do something like this?:
```
func NewRateLimitClients(ctx context.Context, rcFactory factory.ResourceClientFactory) (RateLimitConfigClient, reporter.ReporterResourceClient, error) {
	logf.SetLogger(zap.New())
```

It is also important to note that k8s libraries use the k8s version of zap which does not appear to be interchangeable with the uber zap logger we are using elsewhere, so we can not use `logf.SetLogger(contextutils.LoggerFrom(ctx))`

**Review by elcasteel — APPROVED** [review]

**Comment by sam-heilbron** [comment]
> RateLimitConfigClients is where we hit this because the other "getClient" calls are to modules that use solo kit instead of importing the `controller-runtime` client directly.
> 
> Are you suggesting we do something like this, which seems to also work to supress the messages?:
> 
> ```
> func NewRateLimitClients(ctx context.Context, rcFactory factory.ResourceClientFactory) (RateLimitConfigClient, reporter.ReporterResourceClient, error) {
> 	logf.SetLogger(zap.New())
> ```
> 
> It is also important to note that k8s libraries use the k8s version of zap which does not appear to be interchangeable with the uber zap logger we are using elsewhere, so we can not use `logf.SetLogger(contextutils.LoggerFrom(ctx))`

Yeah what I'm suggesting is that since we're hitting this in a particular area of the code, I imagined we would need to either:
- Fix this in the code itself OR
- Fix this in all kube tests because the same code is run for each test suite

I see other projects have experienced similar issues (https://github.com/kubernetes-sigs/gateway-api/issues/2090), do we know why the error is a quite warning here, but in gloo-fed we were seeing panics?

Have you checked out the mesh repo, or synced with the platform team to see how they configure the controller runtime logger? I imagine they have hit this (or will).

All in all, Im good with the changes as is, and we can merge them, but I think there is a bit more work to be done to either fix the code itself or more of the tests

**Review by sam-heilbron — APPROVED** [review]

File: test/kube2e/gloo/gloo_suite_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sam-heilbron) [reply test/kube2e/gloo/gloo_suite_test.go:56]: Slack thread with more details on this: https://solo-io-corp.slack.com/archives/C03MFATU265/p1691501216502009

We found that this suite has tests which run our SetupFunc directly, and those tests in particular trigger this case. That is why this suite alone needs this change

**Pull request merged**